src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/BlazorDebug/Events.ts

```diff
@@ -0,0 +1,40 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as psList from 'ps-list';
+import { DebugSession } from 'vscode';
+
+import { RazorLogger } from '../RazorLogger';
+
+export async function onDidTerminateDebugSession(
+  event: DebugSession,
+  logger: RazorLogger,
+) {
+  logger.logVerbose('Terminating debugging session...');
+  if (process.platform !== 'win32') {
+    try {
+      const processes = await psList();
+      const devserver = processes.find(
+        (proc: psList.ProcessDescriptor) =>
+          !!(proc && proc.cmd && proc.cmd.includes('blazor-devserver')),
+      );
+      if (devserver) {
+        try {
+          process.kill(devserver.pid);
+          processes.map((proc) => {
+            if (process.ppid === devserver.pid) {
+              process.kill(proc.pid);
```

Comment:
You might want to be a bit more choosy about what processes you kill  -- process ids aren't unique, and will get reused. I have seen many fun bugs because of this. If you have a known list of processes that you know you want to kill - great. Otherwise you could check the process start time (make sure it is after the parent).

Reply:
I did this and moved the `onTerminateDebugSession` handler to inside the configuration provider.

```js
const app = spawn("dotnet", ["run"]);
this.vscodeType.debug.onDidTerminateDebugSession(event => onDidTerminateDebugSession(event, this.logger, app));
```

Is registering event handlers for the debugger in the config provider like this a sensible pattern?

Reply:
That sounds reasonable to me, though do you need to unregister after the debug session ends?

Reply:
Oh, and I should add on the 'choosing processes to kill' front - it looks like the process list API doesn't give start time, but if it is preserving order from the order returned from the system calls, that is good enough.

Reply:
Great point to bring up. This was originally only killing the development server process but I expanded it to kill the entire process tree after seeing [this](https://github.com/microsoft/vscode-js-debug/blob/master/src/targets/node/killTree.ts). I'll use the approach of checking the start-time to make this a little more resilient.